keps/77-dynamically-sized-jobs/README.md

```diff
@@ -0,0 +1,255 @@
+
+To properly support RayCluster, Kueue needs to gracefully handle the scale up and scale down of RayCluster nodes. Concretely this means that scaling the resources used by a RayCluster is appropriately reflected in the respective ClusterQueue without needing to suspend the entire RayCluster.
+
+### Goals
+
+- Gracefully handle resize operations for Kueue jobs (i.e. update quota usage without suspend, enqueue scale ups)
+- Autoscaling RayCluster works with Kueue since it has an autoscaler and there is high demand (MVP)s
+
+### Non-Goals
+
+- Vertical scaling of workloads – Kueue will only handle resize operations that scale Pods horizontally 
+- Support resize for other Kueue jobs such as Job, JobSet, etc (future)
+- Resizing of RayJobs
+- Partial Preemption
+
+## Proposal
+
+Update the Job framework reconciler and introduce new controllers to orchestrate dynamic resizing of jobs. We are only interested in horizontal scaling of jobs (e.g. scaling more replicas). At a high level, this will be accomplished by:
+- Creating Workload Slice objects that represent incremental scale up of jobs. This gives us per-replica control of workload admission. Workload Slices will be garbage collected and consolidated with their parent workloads after successful admission.
+- Adding default scheduling gates to control the scheduling of new pods based on their admission status.
+- Dynamically adjust quotas in ClusterQueue based on scaling events.
+
+For the MVP, we will only focus on admission of **RayClusters** with autoscaling enabled.
+
+### User Stories
+
+#### Story 1 - RayCluster w/ autoscaling
+
+1. The user creates a RayCluster with `enableInTreeAutoscaling: true`.
+2. Kueue admits the RayCluster based on the requested resources in the head pod and worker pod. 
+3. Given the deman of the job, the Ray Autoscaler adjusts the replicas field as it adds or removes Pods from the cluster.
+4. Kueue will not suspend and requeue the RayCluster, instead it will dynamically update ClusterQueue usage based on the scaling event. 
+
+### Notes/Constraints/Caveats (Optional)
+
+If Kueue needs to preempt the resized RayCluster, it would preempt it as a whole, regardless of whether the RayCluster was previously scaled up.
+
+## Design Details
+
+### Workload Slices
+To support horizontal scaling of jobs, we will introduce the concept of a "Workload Slice”. A Workload Slice is a Workload object with an owner reference to the original Workload for a job. Workload Slices represent per-replica changes to a job that were not initially accounted for when the job was created.
+
+The benefit of Workload Slices is that Kueue can evaluate admission on a per-replica basis without changing the existing semantics of the Workload API. Once a Workload Slice is admitted, it will be garbage collected and its resources will be aggregated into the admission status of the parent workload.
+
+
+### Creating Workload Slices
```

Comment:
What if a RayCluster changes its size again before the previously created WorkloadSlice is admitted? Will we create an additional delta WorkloadSlice (say +5 and +3) or the existing WorkloadSlice will be modified?

Reply:
> Will we create an additional delta WorkloadSlice (say +5 and +3) or the existing WorkloadSlice will be modified?

Additional deltas will be created, but there will be some logic added to deleted excess WorkloadSlice objects (either in existing controller or a new one)

Reply:
Why not updating the already existing workload slice for scale ups ? For scale downs updates will have to happen anyway.

Reply:
My understanding is that scale ups beyond the currently admitted workload, have to go through queueing and quota reservation, and updating the existing workload slice might play against the sorting policy?

Reply:
I'm talking about non-admitted workload slice, not the "main" already-admitted workload with which the slice is merged once accepted. It the slice is not yet admitted I don't see a big harm in allowing it to increase if we allow to decrease it.

Reply:
Ah right. It would be consistent with how updates to pending workloads are handled today. AFAIK workloads are updated in-place in that case when compute requests for their jobs are increased. As long as we are sure it'd be correct w.r.t. sorting / queueing policies, it looks like it'd be a good simplification.

Reply:
There is a benefit to having multiple slices: Kueue can admit one of the slices at a time, which could be handy in an already busy cluster.